File: licence_informatique/l3_informatique/technologies_orientee_objet/currency_java/src/main/java/com/franckbarbier/currency/Currency.java
package com.franckbarbier.currency;

interface Obsolete_currency {

    Currency Default_substitute = new Euro();

    default Currency substitute() {
        return Default_substitute;
    }

    default boolean still_active_() {
        return false;
    }

    java.time.LocalDate substitution_date();
}

abstract public class Currency {

    protected final String _common_symbol; // $
    protected final int _iso_code; // 840
    protected final String _iso_symbol; // USD

    protected Currency(String common_symbol, int iso_code, String iso_symbol) {
        _common_symbol = common_symbol;
        _iso_code = iso_code;
        _iso_symbol = iso_symbol;
    }

    public double rate(Currency currency) {
        return 1.D / currency.to_euro_rate();
    }

    abstract public double to_dollar_rate();

    public double convert_to_dollar(final double amount) {
        return amount * to_dollar_rate();
    }

    abstract public double to_euro_rate();

    public double convert_to_euro(final double amount) {
        return amount * to_euro_rate();
    }

}

File: licence_informatique/l3_informatique/conception_application_internet/currency_ts/Pure_JavaScript/Currency.js
class Obsolete_currency {

    static Default_substitute() {
        return new Euro();
    }

    substitute() {
        return Obsolete_currency.Default_substitute();
    }

    still_active_() {
        return false;
    }

    substitution_date() {
        throw new TypeError("abstract");
    }
}

class Currency {
    constructor(common_symbol, iso_code, iso_symbol) {
// Exemple du dollar américain :
        this._common_symbol = common_symbol; // $
        this._iso_code = iso_code; // 840
        this._iso_symbol = iso_symbol; // USD
    }

    rate(currency) {
        return 1. / currency.to_euro_rate();
    }

    to_dollar_rate() {
        throw new TypeError("abstract");
    }

    convert_to_dollar(amount) {
        return amount * this.to_dollar_rate();
    }

    to_euro_rate() {
        throw new TypeError("abstract");
    }

    convert_to_euro(amount) {
        return amount * this.to_euro_rate();
    }
}

class Euro extends Currency {

    constructor() {
        super("€", 978, "EUR");
    }

    to_dollar_rate() {
        return 1.1; // Taux pris le 17 sept. 2019
    }

    to_euro_rate() {
        return 1.;
    }
}

class Lats extends Currency { // Monnaie de la Lettonie remplacée par l'Euro le 1er janvier 2014 

    static Substitution_date() {
        return Date.parse('01 Jan 2014 00:00:00 GMT');
    }

    constructor(obsolete_currency = new Obsolete_currency) { // There is *NO* multiple class/interface inheritance/implementation in JavaScript!
        super("Ls", 428, "LVL");
        this._obsolete_currency = obsolete_currency;
    }

    to_dollar_rate() {
        const substitute = this._obsolete_currency.substitute();
        return this.rate(substitute) * substitute.to_dollar_rate();
    }

    to_euro_rate() {
        return 1. / 0.702804; // Taux pris le 1er janvier 2014
    }

    substitution_date() {
        return Lats.Substitution_date().toString();
    }
}

const lats = new Lats();
console.log(lats.convert_to_euro(100));
console.log(lats.convert_to_dollar(100));
